### PR TITLE
GraalSOM - a Truffle Graal implementation of som-java

### PR DESCRIPTION
**Pull request by sophie-kaleba**
Register som-java as a Truffle language and have basic interpretation (i.e. not taking advantage of truffle/graal optimisation yet) working. 

Still a WIP - I assume there is a lot of redundancy in the code, and requires a proper cleanup (notably, a better handling for `Source` objects and proper `Context` initialization from GraalSOMLanguage)

@smarr could you please have a look ?

### CODE REVIEW

File: build.xml [hunk 1]
```diff
@@ -1,126 +1,417 @@
...
     <property name="classes.dir" value="${build.dir}/classes"/>
-    <property name="junit.version" value="4.12" />
 
-    <property name="checkstyle.version" value="8.14" />
+    <property name="mx.cmd" value="../../mx/mx" />
 
     <property environment="env"/>
 
-    <path id="project.classpath">
+<!--    <property name="src.dir"     value="src"/>-->
```
Comment (smarr) [inline build.xml:67]: If you don't need those, could we remove them?

File: build.xml [hunk 2]
```diff
@@ -1,126 +1,417 @@
...
+<!--    <property name="test.dir"    value="tests"/>-->
+<!--    <property name="lib.dir"     value="libs" />-->
+<!--    <property name="build.dir"   value="build"/>-->
+<!--    <property name="classes.dir" value="${build.dir}/classes"/>-->
+
+
+    <property name="checkstyle.version" value="8.14" />
+
+<!--    <path id="project.classpath">-->
```
Comment (smarr) [inline build.xml:76]: Same here, I'd remove all old code. We have other places where we keep it for reference.

File: build.xml [hunk 3]
```diff
@@ -1,126 +1,417 @@
...
+    </path>
+
+    <path id="common.cp">
         <pathelement location="${classes.dir}" />
-        <pathelement location="${test.dir}" />
+        <pathelement location="${svm.build}/svm-core.jar" />
+        <pathelement location="${tools.build}/truffle-profiler.jar" />
+    </path>
+
+<!--    //TODO-change-->
```
Comment (smarr) [inline build.xml:94]: Change? In which way?

File: build.xml [hunk 4]
```diff
@@ -1,126 +1,417 @@
...
+            <arg value="diff-index" />
+            <arg value="--quiet" />
+            <arg value="--exit-code" />
+            <arg value="--ignore-submodules" />
+            <arg value="HEAD" />
+        </exec>
+        <travis target="eclipseformat-check" />
     </target>
 
+<!--    <target name="clean">-->
```
Comment (smarr) [inline build.xml:242]: Remove?

File: build.xml [hunk 5]
```diff
@@ -1,126 +1,417 @@
...
     <target name="checkstyle-jar">
-        <mkdir dir="${lib.dir}" />
+        <travis target="checkstyle-jar" start="Get Checkstyle Jar" />
         <get src="https://github.com/checkstyle/checkstyle/releases/download/checkstyle-${checkstyle.version}/checkstyle-${checkstyle.version}-all.jar"
              usetimestamp="true"
              dest="${lib.dir}/checkstyle-${checkstyle.version}-all.jar" />
+        <travis target="checkstyle-jar" />
     </target>
-    
-    <target name="checkstyle" depends="checkstyle-jar" description="Check Code with Checkstyle">
```
Comment (smarr) [inline build.xml:75]: Is there still a check style target?
Would be good to keep this for CI.

File: build.xml [hunk 6]
```diff
@@ -1,126 +1,417 @@
...
             <formatter type="brief" usefile="false"/>
         </junit>
 
-        <java classname="som.vm.Universe" fork="true" failonerror="true">
+        <java classname="som.Launcher" fork="true" failonerror="true">
             <classpath refid="project.classpath" />
             <jvmarg value="-ea" />
             <jvmarg value="-esa" />
-            <arg line="-cp Smalltalk TestSuite/TestHarness.som" />
+            <arg line="-cp Smalltalk TestSuite/TestHarness.som  --ignore-inefficacies" />
```
Comment (smarr) [inline build.xml:399]: Hm, is this necessary? Do you have things that fail without it?

File: build.xml [hunk 7]
```diff
@@ -1,126 +1,417 @@
...
         </java>
         <fail message="Basic tests failed." if="test.failed" />
     </target>
 
-    <target name="test" depends="unit-tests">
+    <target name="som-test" depends="compile" description="Test som script">
+        <travis target="som-test" start="SOM Test" />
+
+        <exec executable="./som" failonerror="true">
+            <arg value="TestSuite/TestHarness.som" />
```
Comment (smarr) [inline build.xml:408]: Here you don't seem to need the `--ignore-inefficacies` option.
There are two classes of assertions in unit tests: things that need to work, and things that should work, but not all implementations can provide it reliably.
Inefficacies are what I chose to call the cases that are not strictly required to pass.

File: som [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+import shlex
+
+BASE_DIR    = os.path.dirname(os.path.realpath(__file__))
+TRUFFLE_DIR = BASE_DIR + '/libs/truffle'
+JAVA_HOME   = os.getenv('JAVA_HOME',  None)
+JVMCI_BIN   = os.getenv('JVMCI_BIN',  None)
+GRAAL_HOME  = os.getenv('GRAAL_HOME',  None)
+GRAAL_FLAGS = os.getenv('GRAAL_FLAGS', None)
+
+# GRAAL_LOCATIONS = ['~/.local/graal-core/',
```
Comment (smarr) [inline som:14]: Old, let's hope we can remove this

File: som [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+if args.print_graal_options:
+    flags += ['-XX:+JVMCIPrintProperties']
+
+if args.java_properties:
+    flags += ['-D' + property for property in args.java_properties]
+
+if args.java_args:
+    JAVA_ARGS += ['-' + property for property in args.java_args]
+
+#flags += ['-Dsom.tools=' + BASE_DIR + '/tools']
```
Comment (smarr) [inline som:279]: Would remove

File: src/som/GraalSOMLanguage.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        StandardTags.RootBodyTag.class,
+        StandardTags.ExpressionTag.class,
+        DebuggerTags.AlwaysHalt.class})
+public final class GraalSOMLanguage extends TruffleLanguage<Universe> {
+
+    public static final String ID = "GS";
+    public static final String MIME_TYPE = "application/x-graal-som";
+    public static String[] args;
+
+    @Option(help = "Help Text.", category = OptionCategory.USER) public static final OptionKey<String> TestSelector = new OptionKey<>("");
```
Comment (smarr) [inline src/som/GraalSOMLanguage.java:38]: Could we have a bit of help text here?
And if it's just for documentation.

File: src/som/GraalSOMLanguage.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    protected Universe createContext(final Env env) {
+        return new Universe(env, this);
+    }
+
+    @Override
+    protected void initializeContext(final Universe context) {
+        Env currentEnv = context.getEnv();
```
Comment (smarr) [inline src/som/GraalSOMLanguage.java:52]: For simplicity, and to stay as close as possible to the other SOM implementations, I would use the same approach as SOMns. Simply letting the language take care of the various bits itself, and ignore the truffle framework.

Currently, I don't think we plan to look into language interop. So, I think it's perfectly fine to take the simple way out.

File: src/som/GraalSOMLanguage.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    @Override
+    protected OptionDescriptors getOptionDescriptors() {
+        // this class is generated by the annotation processor
+        return new GraalSOMLanguageOptionDescriptors();
+    }
+
+    @Override
+    protected boolean isObjectOfLanguage(final Object object) {
+        return object instanceof SAbstractObject;
+        //TODO - should throw an exception
```
Comment (smarr) [inline src/som/GraalSOMLanguage.java:74]: Why should it throw?

File: src/som/GraalSOMLanguage.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    public static Universe getCurrentContext() {
+        return getCurrentContext(GraalSOMLanguage.class);
+    }
+
+    @Override
+    protected CallTarget parse(final ParsingRequest request) throws Exception {
+        Universe context = getCurrentContext();
```
Comment (smarr) [inline src/som/GraalSOMLanguage.java:83]: I would suggest to keep naming it universe, just so the code is similar to the other implementation.
Also, to help me reviewing. With `universe`, I know immediately what it is :)

```suggestion
        Universe universe = getCurrentContext();
```

File: src/som/GraalSOMLanguage.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        return Truffle.getRuntime().createCallTarget(new RootNode(this) {
+
+            @Override
+            public Object execute(final VirtualFrame frame) {
+                try {
+                    if (context.isForTesting()) {
+                        return context.interpret(args, context.testClass(), context.testSelector());
+                    }
+                    else return context.interpret(args);
+                    //TODO - leave the object system initialisation in the initializeContext method, along the lines of...
```
Comment (smarr) [inline src/som/GraalSOMLanguage.java:93]: Hm, interesting design question. What are the drawbacks/benefits? Is this related to the earlier comment I made? Perhaps this solution here is fine?

I guess a big question is whether we want to be able to call multiple times `eval()` on the same `engine`?

File: src/som/vm/Universe.java [hunk 15]
```diff
@@ -53,21 +58,40 @@
 import som.vmobjects.SSymbol;
 
 
-public class Universe {
+public final class Universe {
 
-  public static void main(final String[] arguments) {
-    // Create Universe
-    Universe u = new Universe();
+  static final Layout LAYOUT = Layout.createLayout();
```
Comment (smarr) [inline src/som/vm/Universe.java:63]: Hm, I wouldn't start with things like object layouts just yet.
Let's see that we get the basic execution mechanisms ready for the partial evaluation.

File: src/som/vm/Universe.java [hunk 16]
```diff
@@ -53,21 +58,40 @@
...
-    // Start interpretation
-    try {
-      u.interpret(arguments);
-    } catch (ProgramDefinitionError e) {
-      u.errorExit(e.toString());
-    }
+  private final TruffleLanguage.Env env;
+  private final GraalSOMLanguage language;
+  private final String testClass;
+  private final String testClasspath;
```
Comment (smarr) [inline src/som/vm/Universe.java:68]: Ok, so, I think this relates to the question of whether to allow multiple `eval()` calls from the launcher.
A Universe needs to initialize the object system, so, that takes loading and parsing various classes, which takes time. So, I'd try to not throw it away unnecessarily.

File: src/som/vmobjects/SClass.java [hunk 17]
```diff
@@ -302,4 +307,18 @@ public String toString() {
...
   // Static field indices and number of class fields
   static final int numberOfClassFields = numberOfObjectFields;
+
+  /**
+   * INTEROP
+   * A cast of a SClass to a java String will return the name of the class 
+   */
+  @ExportMessage
+  public boolean isString() {
+    return true;
```
Comment (smarr) [inline src/som/vmobjects/SClass.java:317]: Hm, this feels wrong. We can't simply do the same as TruffleSOM, and get the Java object out directly?

File: src/som/vmobjects/SDouble.java [hunk 18]
```diff
@@ -134,4 +139,92 @@ public SObject primLessThan(final SNumber right, final Universe universe) {
...
+   */
+
+  @ExportMessage
+  boolean isNumber() {
+    return true;
+  }
+
+  @ExportMessage
+  boolean fitsInByte() {
+    return false;
```
Comment (smarr) [inline src/som/vmobjects/SDouble.java:155]: Do all of these have to be implemented?
Or would the double one suffice?

File: src/som/vmobjects/SDouble.java [hunk 19]
```diff
@@ -134,4 +139,92 @@ public SObject primLessThan(final SNumber right, final Universe universe) {
...
+  boolean fitsInInt() {
+    return false;
+  }
+
+  @ExportMessage
+  boolean fitsInDouble() {
+    return true;
+  }
+
+  @ExportMessage byte asByte() throws UnsupportedMessageException {
```
Comment (smarr) [inline src/som/vmobjects/SDouble.java:183]: Same here, can we get rid of all but the double version?

File: src/som/vmobjects/SInteger.java [hunk 20]
```diff
@@ -286,4 +291,96 @@ public SObject primLessThan(final SNumber right, final Universe universe) {
...
+    return false;
+  }
+
+  @ExportMessage
+  boolean fitsInFloat() {
+    return false;
+  }
+
+  @ExportMessage
+  boolean fitsInLong() {
```
Comment (smarr) [inline src/som/vmobjects/SInteger.java:323]: Same here, is it ok to have just the long stuff?

File: build.xml (hunk 1 above)
Comment (sophie-kaleba) [reply build.xml]: Yes, this and the other commented bits were a remainder of a first cleanup attempt. They can be removed indeed!

File: build.xml (hunk 5 above)
Comment (sophie-kaleba) [reply build.xml:75]: It was missing indeed, thanks!

File: build.xml (hunk 6 above)
Comment (sophie-kaleba) [reply build.xml]: It does not fail without it, and I don't even remember why I put this here in the first place. I removed it.

File: build.xml (hunk 3 above)
Comment (sophie-kaleba) [reply build.xml]: The change mentioned was added earlier on but I forgot to remove this comment. Can be removed

File: src/som/vmobjects/SDouble.java (hunk 18 above)
Comment (sophie-kaleba) [reply src/som/vmobjects/SDouble.java:155]: Hmm, from my understanding (and from the compiler complaints), I must implement all of those as they turn abstract if specify that `SDouble` should be considered as a Number. 
For instance, this is the code for `fitsInByte` in the `InteropLibrary` class:

```
    @Abstract(ifExported = "isNumber")
    public boolean fitsInByte(Object receiver) {
        return false;
    }
```

The library does not offer to chose a type more specific than Number.

Comment (smarr) [reply src/som/vmobjects/SDouble.java:155]: Hm, sad. Though, could you move the common parts into `SNumber` perhaps?

Comment (sophie-kaleba) [reply src/som/vmobjects/SDouble.java:155]: Yes, I was unsure how the annotations would work with subclasses but it seems to work fine. I'll modify SNumber, SDouble and SInteger accordingly. Thanks!

File: src/som/vmobjects/SClass.java (hunk 17 above)
Comment (sophie-kaleba) [reply src/som/vmobjects/SClass.java]: Indeed, I went back to the TruffleSOM way of handling the java object

File: src/som/Launcher.java [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+        PrintStream err = System.err;
+
+        Context.Builder builder = Context.newBuilder();
+        builder.arguments(GraalSOM, arguments);
+
+        // Try to build the context in which we will execute the source code
+        try {
+            context = builder.build();
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (sophie-kaleba) [reply src/som/Launcher.java]: Actually, not. I removed this try catch

File: src/som/Launcher.java [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+
+    private static final String GraalSOM = "GS";
+
+    public static void main(final String[] arguments) throws IOException {
+      String file = null;
+      Source source;
+
+      //TODO - this isn't right - ignored at the moment
+      file = "core-lib/Examples/Echo.som";
+      source = Source.newBuilder(GraalSOM, new File(file)).internal(true).buildLiteral();
```
[the remark this answers is not in the document]
Comment (sophie-kaleba) [reply src/som/Launcher.java]: Done

File: src/som/GraalSOMLanguage.java (hunk 14 above)
Comment (sophie-kaleba) [reply src/som/GraalSOMLanguage.java]: For now, let's just support one `eval()` on the same engine and get rid of the (yet) unnecessary complexity

File: src/som/GraalSOMLanguage.java (hunk 12 above)
Comment (sophie-kaleba) [reply src/som/GraalSOMLanguage.java]: I think the initial intent was to get an explicit exception message out of it, but I was not entirely sure how and what for the framework uses this method. From what I have seen, it may be wiser to let the framework handling the result.

File: src/som/GraalSOMLanguage.java (hunk 11 above)
Comment (sophie-kaleba) [reply src/som/GraalSOMLanguage.java]: Yes, I removed that part and the subsequent commented parts in the code about this matter. Let's keep to one `eval` per engine for now.

**Comment by smarr** [comment]
Once you're happy with the last few changes I pushed, I think we can merge this.

**Comment by sophie-kaleba** [comment]
I am ok with the last changes, merging into graalsom branch. Thank you!
